### ISSUE TITLE
iptables firewall bouncer starting before the docker daemon on boot?

### ISSUE THREAD

**Issue by dakota-marshall** [crowdsecurity/cs-firewall-bouncer#216]
Based on my logs, it looks like by default the iptables firewall bouncer tries to initialize the iptables rules before the docker daemon can create its chain when the server is starting. Here is the error I got on boot today:

```
time="12-11-2022 06:43:29" level=info msg="iptables set-up : /usr/sbin/iptables -I DOCKER-USER -m set --match-set crowdsec-blacklists src -j DROP"
time="12-11-2022 06:43:29" level=warning msg="Error inserting set in iptables (/usr/sbin/iptables -I DOCKER-USER -m set --match-set crowdsec-blacklists src -j DROP): exit status 1 : iptables: No chain/target/match by that name.\n"
time="12-11-2022 06:43:29" level=fatal msg="iptables init failed: while inserting set in iptables: exit status 1"
```

Here is my bouncer config:

```yaml
mode: iptables
pid_dir: /var/run/
update_frequency: 10s
daemonize: true
log_mode: file
log_dir: /var/log/
log_level: info
log_compression: true
log_max_size: 100
log_max_backups: 3
log_max_age: 30
api_url: http://127.0.0.1:8080/
api_key:
insecure_skip_verify: false
disable_ipv6: true
deny_action: DROP
deny_log: false
supported_decisions_types:
  - ban
#to change log prefix
#deny_log_prefix: "crowdsec: "
#to change the blacklists name
blacklists_ipv4: crowdsec-blacklists
blacklists_ipv6: crowdsec6-blacklists
#if present, insert rule in those chains
iptables_chains:
#  - INPUT
#  - FORWARD
  - DOCKER-USER

## nftables
nftables:
  ipv4:
    enabled: true
    set-only: false
    table: crowdsec
    chain: crowdsec-chain
  ipv6:
    enabled: true
    set-only: false
    table: crowdsec6
    chain: crowdsec6-chain
# packet filter
pf:
  # an empty string disables the anchor
  anchor_name: ""
```

I haven't made any changes to either the docker systemd service or the bouncer service, but just for giggles here is what my service file looks like:

```systemd
[Unit]
Description=The firewall bouncer for CrowdSec
After=syslog.target network.target remote-fs.target nss-lookup.target crowdsec.service
Before=netfilter-persistent.service

[Service]
Type=notify
ExecStart=/usr/sbin/crowdsec-firewall-bouncer -c /etc/crowdsec/bouncers/crowdsec-firewall-bouncer.yaml
ExecStartPre=/usr/sbin/crowdsec-firewall-bouncer -c /etc/crowdsec/bouncers/crowdsec-firewall-bouncer.yaml -t
ExecStartPost=/bin/sleep 0.1

[Install]
WantedBy=multi-user.target

```
I have not tried adding docker.service to the After section of the unit file yet, as this is a production system. This is Ubuntu 20.04.5 if that matters either.

Any advice? Let me know if I may just be missing something

**Comment by LaurenceJJones** [comment]
rather odd, is docker listed to enable on startup via systemd?
If so could you add to the after line
```
After=docker.service
```

**Comment by StSixtus** [comment]
Not working in my case.
I added the After=docker.service at the end of the After line
After=syslog.target network.target remote-fs.target nss-lookup.target crowdsec.service docker.service
 and ReloadPropagatedFrom=docker.service to /etc/systemd/system/crowdsec-firewall-bouncer.service 
But it isn't working 
Even worse the /var/log/crowdsec-firewall-bouncer.log and systemctl status crowdsec-firewall-bouncer disagrees. The logs are saying the process failed, while the service status is fine, despite the Main PID has exited.

Otherwise, if I restart the service manually after some time, it works

Here are the "conflicting" results

 cat /var/log/crowdsec-firewall-bouncer.log
time="29-11-2022 23:51:32" level=info msg="backend type : nftables"
time="29-11-2022 23:51:32" level=info msg="nftables initiated"
time="29-11-2022 23:51:32" level=info msg="Using API key auth"
time="29-11-2022 23:51:32" level=info msg="Processing new and deleted decisions . . ."
time="29-11-2022 23:52:02" level=error msg="auth-api: auth with api key failed return nil response, error: dial tcp 172.23.0.3:8080: i/o timeout"
time="29-11-2022 23:52:02" level=error msg="Get \"http://172.23.0.3:8080/v1/decisions/stream?startup=true\": dial tcp 172.23.0.3:8080: i/o timeout"
time="29-11-2022 23:52:02" level=error msg="terminating bouncer process"
time="29-11-2022 23:52:02" level=error msg="process return with error: stream api init failed"
time="29-11-2022 23:52:02" level=info msg="removing 'crowdsec' table"
time="29-11-2022 23:52:02" level=info msg="removing 'crowdsec6' table"
root@HVDebian:~# systemctl status crowdsec-firewall-bouncer
● crowdsec-firewall-bouncer.service - The firewall bouncer for CrowdSec
     Loaded: loaded (/etc/systemd/system/crowdsec-firewall-bouncer.service; enabled; vendor preset: enabled)
     Active: inactive (dead) since Tue 2022-11-29 23:52:02 CET; 3min 8s ago
    Process: 2016 ExecStartPre=/usr/sbin/crowdsec-firewall-bouncer -c /etc/crowdsec/bouncers/crowdsec-firewall-bouncer.yaml -t (code=e>
    Process: 2043 ExecStart=/usr/sbin/crowdsec-firewall-bouncer -c /etc/crowdsec/bouncers/crowdsec-firewall-bouncer.yaml (code=exited,>
    Process: 2058 ExecStartPost=/bin/sleep 0.1 (code=exited, status=0/SUCCESS)
   Main PID: 2043 (code=exited, status=0/SUCCESS)
        CPU: 29ms

nov. 29 23:51:32 HVDebian systemd[1]: Starting The firewall bouncer for CrowdSec...
nov. 29 23:51:32 HVDebian crowdsec-firewall-bouncer[2016]: time="2022-11-29T23:51:32+01:00" level=info msg="crowdsec-firewall-bouncer >
nov. 29 23:51:32 HVDebian crowdsec-firewall-bouncer[2016]: time="2022-11-29T23:51:32+01:00" level=info msg="config is valid"
nov. 29 23:51:32 HVDebian crowdsec-firewall-bouncer[2043]: time="2022-11-29T23:51:32+01:00" level=info msg="crowdsec-firewall-bouncer >
nov. 29 23:51:32 HVDebian systemd[1]: Started The firewall bouncer for CrowdSec.
nov. 29 23:52:02 HVDebian systemd[1]: crowdsec-firewall-bouncer.service: Succeeded.

**Comment by LaurenceJJones** [comment]
Seems your error is happening because crowdsec is inside a container? (presumed from 172 address?) and on startup if the bouncer fails to reconnect it dies....
```
time="29-11-2022 23:52:02" level=error msg="auth-api: auth with api key failed return nil response, error: dial tcp 172.23.0.3:8080: i/o timeout"
time="29-11-2022 23:52:02" level=error msg="Get "[http://172.23.0.3:8080/v1/decisions/stream?startup=true\](http://172.23.0.3:8080/v1/decisions/stream?startup=true%5C)": dial tcp 172.23.0.3:8080: i/o timeout"
```

**Comment by StSixtus** [comment]
You are right, my crowdsec is in a docker container, and when the firewall-bouncer starts it cannot connect to crowdsec-agent (for some reason (probably delay) and the bouncer shutdowns, leaving my host with no protection :-(
I applied the suggested workaround, without success, so I commented this issue.

As far I understand, the FW starting before the docker daemon on boot is the case of this issue. I just added that the bouncer should not start before the crowdsec container inside docker. Does it deserve another issue ?

And another issue should probably be opended, for the misleading systemctl status of the firewall-bouncer service. Others seem to comment in this way https://github.com/crowdsecurity/cs-firewall-bouncer/issues/217 "But I'm wanted to underline that the bouncer does not resist to a network failure at boot time (whatever the reason).
I my mind, the bouncer should be designed to resist to any kind of storm."

**Comment by floari** [comment]
I am facing the same issue :)
I did a workaround by restarting the bouncer after 2minutes via cron: @reboot sleep 120 && systemctl start crowdsec-firewall-bouncer

This is not optimal. Are there other options?

**Comment by StSixtus** [comment]
@floari 

I've done a restart too, but only in the service description file. Moreover, it garantees a coherent status with
`systemctl status crowdsec-firewall-bouncer`

I added options, so the service cannot be active with an exited process, and that it must restart if inactive/exited.

RemainAfterExit=no
Restart=always
RestartSec=20

You can find the location of the service file with the command `systemctl status crowdsec-firewall-bouncer`
Here is my complete `/etc/systemd/system/crowdsec-firewall-bouncer.service`

```
[Unit]
Description=The firewall bouncer for CrowdSec
After=syslog.target network.target remote-fs.target nss-lookup.target crowdsec.service docker.service
Before=netfilter-persistent.service
ReloadPropagatedFrom=docker.service
StartLimitIntervalSec=20

[Service]
Type=notify
RemainAfterExit=no
Restart=always
RestartSec=20
ExecStart=/usr/sbin/crowdsec-firewall-bouncer -c /etc/crowdsec/bouncers/crowdsec-firewall-bouncer.yaml
ExecStartPre=/usr/sbin/crowdsec-firewall-bouncer -c /etc/crowdsec/bouncers/crowdsec-firewall-bouncer.yaml -t
ExecStartPost=/bin/sleep 0.1

[Install]
WantedBy=multi-user.target


```

**Comment by floari** [comment]
@StSixtus 

Thanks - thats working perfectly.
Maybe something to add in the documentation!
